lib/ansible/module_utils/redfish_utils.py

```diff
@@ -283,7 +298,7 @@ def get_storage_controller_inventory(self):
         properties = ['Name', 'Status']
 
         # Find Storage service
-        response = self.get_request(self.root_uri + self.systems_uri)
+        response = self.get_request(self.root_uri + self.systems_uris)
```

Comment:
I would leave `get_storage_controller_inventory` function largely intact in this case, since it already sort-of works for single system. The only thing that I would change is to replace `self.systems_uri` with an explicit function argument to decouple it from the global state.

Then I would introduce another function like `get_multi_storage_controller_inventory` that just loop over the `self.systems_uris`, calls `get_storage_controller_inventory` and merges results.

How to merge results is something that maintainers should guide you through since this will also affect the final result format and maintainers are the ones who should define that.

Reply:
@tadeboro thank you :)
@billdodd any idea how to resolve this.

Reply:
@dnuka - Yes, I like the approach that @tadeboro outlined.

Reply:
@tadeboro , I can refactor the  `get_storage_controller_inventory(self)` to `get_storage_controller_inventory(self, system_uri)`. and also all the methods, that access `self.systems_uris` including `get_system_inventory(self)` to `get_system_inventory(self, system_uri)`. But, the problem is, then I'm ended up with `TypeError: get_system_inventory() takes exactly 2 arguments (1 given)`

Reply:
You are missing the part about introducing a wrapper function that will iterate over the array and call. What I had in mind is something similar to this:

```diff
diff --git a/lib/ansible/module_utils/redfish_utils.py b/lib/ansible/module_utils/redfish_utils.py
index adbb3316d2..8e4d43db06 100644
--- a/lib/ansible/module_utils/redfish_utils.py
+++ b/lib/ansible/module_utils/redfish_utils.py
@@ -124,23 +124,23 @@ class RedfishUtils(object):
         response = self.get_request(self.root_uri + uri)
         if response['ret'] is False:
             return response
-        data = response['data']
-        if 'Systems' not in data:
+        if 'Systems' not in response['data']:
             return {'ret': False, 'msg': "Systems resource not found"}
-        else:
-            systems = data["Systems"]["@odata.id"]
-            response = self.get_request(self.root_uri + systems)
-            if response['ret'] is False:
-                return response
-            data = response['data']
-            if data.get(u'Members'):
-                for member in data[u'Members']:
-                    systems_service = member[u'@odata.id']
-                    self.systems_uri = systems_service
-                    return {'ret': True}
-            else:
-                return {'ret': False,
-                        'msg': "ComputerSystem's Members array is either empty or missing"}
+
+        response = self.get_request(self.root_uri + data["Systems"]["@odata.id"])
+        if response['ret'] is False:
+            return response
+
+        self.system_uris = [
+            i["@odata.id"] for i in response["data"].get("Members", [])
+        ]
+        if not self.system_uris:
+            return {
+                'ret': False,
+                'msg': "ComputerSystem's Members array is either empty or missing",
+            }
+
+        return {'ret': True}
 
     def _find_updateservice_resource(self, uri):
         response = self.get_request(self.root_uri + uri)
@@ -275,7 +275,16 @@ class RedfishUtils(object):
                         return response
         return {'ret': True}
 
-    def get_storage_controller_inventory(self):
+    def get_multi_storage_controller_inventory(self):
+        ret = True
+        entries = {}
+        for system_uri in self.system_uris:
+            inventory = self.get_storage_controller_inventory(system_uri)
+            ret = inventory.pop("ret") and ret
+            entries[system_id] = inventory
+        return dict(ret=ret, entries=entries)
+
+    def get_storage_controller_inventory(self, system_uri):
         result = {}
         controller_list = []
         controller_results = []
@@ -283,7 +292,7 @@ class RedfishUtils(object):
         properties = ['Name', 'Status']
 
         # Find Storage service
-        response = self.get_request(self.root_uri + self.systems_uri)
+        response = self.get_request(self.root_uri + system_uri)
         if response['ret'] is False:
             return response
         data = response['data']
diff --git a/lib/ansible/modules/remote_management/redfish/redfish_facts.py b/lib/ansible/modules/remote_management/redfish/redfish_facts.py
index 9e8d088195..50a6d7bae0 100644
--- a/lib/ansible/modules/remote_management/redfish/redfish_facts.py
+++ b/lib/ansible/modules/remote_management/redfish/redfish_facts.py
@@ -218,7 +218,7 @@ def main():
                 elif command == "GetNicInventory":
                     result["nic"] = rf_utils.get_nic_inventory(category)
                 elif command == "GetStorageControllerInventory":
-                    result["storage_controller"] = rf_utils.get_storage_controller_inventory()
+                    result["storage_controller"] = rf_utils.get_multi_storage_controller_inventory()
                 elif command == "GetDiskInventory":
                     result["disk"] = rf_utils.get_disk_inventory()
                 elif command == "GetBiosAttributes":
```

Repeat this for all the functions that access `self.systems_uri` directly and you should be gold.

But before you finalize anything, get some input from the maintainers about the final result structure. What I did here is just an example merge that felt at least reasonably useful.

Reply:
thank you @tadeboro :)